### PR TITLE
using LM, dependency parse, and rule-based methods for neutral generation

### PR DESCRIPTION
**Pull request by tonysun9**

### CODE REVIEW

File: neutral_generation/constants.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            'theirs']   # This pen is his --> This pen is theirs
+}
+
+NON_INJECTIVE_PRONOUNS = {
+    # many-to-one mapping
+    'he': 'they',
+    'she': 'they',
+    "she's": "they're",
+    "he's": "they're",
+    'herself': 'themself',  # TODO: be careful with cases such as "Sarah herself" --> "Sarah themself"
```
Comment (melvinjosej) [inline neutral_generation/constants.py:16]: What's tricky about the "Sarah themself" case? We could refer to Sarah in the gender-neutral way right?

File: neutral_generation/constants.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+# verbs
+RULE_CHANGE_VERBS = {
+    'is': 'are',
+    'has': 'have',
+    'was': 'were'
+}
+
+IRREGULAR_PRESENT_SIMPLE_VERBS = {
+    'is': 'are',
+    "'s": "'re",
```
Comment (melvinjosej) [inline neutral_generation/constants.py:60]: Note that 's could also be a contraction of has. See https://en.wiktionary.org/wiki/he%27s#:~:text=he%20has-,Usage%20notes,however%2C%20use%20he's%20for%20both.

File: neutral_generation/smart_convert.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+nlp = spacy.load("en_core_web_sm")
+
+# Load pre-trained model (weights)
+model = OpenAIGPTLMHeadModel.from_pretrained('openai-gpt')
+model.eval()
+# Load pre-trained model tokenizer (vocabulary)
+tokenizer = OpenAIGPTTokenizer.from_pretrained('openai-gpt')
+
+
+def convert(s):
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:20]: Let's change 's' to something more descriptive. Avoid single character var names.

File: neutral_generation/smart_convert.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    # output: sentence in gender-neutral form
+
+    # pluralize verbs: change all verbs in third-person singular to third-person plural
+    auxiliaries = identify_verbs_and_auxiliaries(s)
+    s = pluralize_verbs(s, auxiliaries)
+
+    # use regex to replace pronouns and gendered terms that have a clear mapping
+    SIMPLE_REPLACE = EASY_PRONOUNS
+    SIMPLE_REPLACE.update(GENDERED_TERMS)
+    for p, r in SIMPLE_REPLACE.items():
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:31]: Again, remove single char vars.

File: neutral_generation/smart_convert.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        s = regex_token_replace(s, p, r)
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:32]: This segment here can be profiled and optimized. Have you tried to see if the regex takes a lot of time?

If so, we can flip it. i.e loop through each token in the input and then in the inner loop go over all possible replacements. That might be faster.

File: neutral_generation/smart_convert.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+    # use a LM to break ties for pronouns when there is a one-to-many mapping
+    for p, choices in NON_FUNCTION_PRONOUNS.items():
+        s = smart_pronoun_replace(s, p, choices)
+
+    return s
+
+
+def regex_token_replace(s, old, new):
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:41]: Specify inputs and outputs for this function and all functions.

File: neutral_generation/smart_convert.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    for j in range(len(replace_map)):
+        pattern = re.compile(r'\b{}\b'.format(replace_map[j][0]))
+        # TODO: add an exception to regex replacement when an apostrophe follows the pronoun
+        # e.g. He'shan is a location, but with this rule it would become They'shan
+        s = re.sub(pattern, replace_map[j][1], s)
+
+    return s
+
+
+def capitalize(word):
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:55]: I think Python has a built-in capitalize function that you can use.
See: https://www.geeksforgeeks.org/string-capitalize-python/

File: neutral_generation/smart_convert.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+
+    # rule_change includes 4 tenses in total:
+    #   the present continuous, the 2 present perfect tenses, and the past continuous
+    rule_change = ['present continuous', 'present perfect', 'past continuous']
+
+    # TODO: categorize tense for "do" verbs
+    verb_tense = dict()
+    for verb, auxiliary_list in auxiliaries.items():
+        auxiliary_text = ' '.join(auxiliary.text for auxiliary in auxiliary_list)
+        full_verb = auxiliary_text + ' ' + verb.text
```
Comment (melvinjosej) [inline neutral_generation/smart_convert.py:159]: String concat using '+' in python is inefficient. If you want to optimize your code, consider alternatives.
See: https://blog.mclemon.io/python-efficient-string-concatenation-in-python-2016-edition

**Review by melvinjosej — COMMENTED** [review]
Sending some initial comments. Will send in more comments in a bit.

File: neutral_generation/constants.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# pronouns
+NON_FUNCTION_PRONOUNS = {   # need to change pronoun based on context
```
Comment (kelliemwebster) [inline neutral_generation/constants.py:2]: Optional.

These pronoun mappings actually become unique when you have the dependency structure, e.g. "her" when possessive becomes "their", while "her" when an object becomes "them". While checking the tags introduces some code complexity, it does remove the dependency on an LM. 

See "poss" and "*obj" in https://spacy.io/api/annotation#dependency-parsing

File: neutral_generation/constants.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    'was': 'were'
+}
+
+IRREGULAR_PRESENT_SIMPLE_VERBS = {
+    'is': 'are',
+    "'s": "'re",
+    'has': 'have',
+    'does': 'do',
+    'doesn': 'don',
+    'goes': 'go'
```
Comment (kelliemwebster) [inline neutral_generation/constants.py:64]: I trust this is based on the tokenization you've seen in SpaCy but just wanted to call out that it's strange. Typically, we tokenize "doesn't" as "does" + "n't"

File: neutral_generation/smart_convert.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+
+
+def regex_token_replace(s, old, new):
+    # replace occurrences of a token in a string with a new token
+    # TODO: can look into redoing replace_map with SpaCy (they save capitalization as well)
+    replace_map = [[old, new], [capitalize(old), capitalize(new)], [old.upper(), new.upper()]]
+
+    for j in range(len(replace_map)):
+        pattern = re.compile(r'\b{}\b'.format(replace_map[j][0]))
+        # TODO: add an exception to regex replacement when an apostrophe follows the pronoun
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:48]: You can check that the token is POS tagged as a pronoun to avoid errors like the one you describe here.

See "PRP" / "PRP$" in https://spacy.io/api/annotation#pos-tagging

File: neutral_generation/constants.py (hunk 1 above)
Comment (kelliemwebster) [reply neutral_generation/constants.py:16]: Separate question -- do we want to loop in someone with expertise in gender neutral reference?

File: neutral_generation/smart_convert.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+def score(sentence):
+    tokenize_input = tokenizer.tokenize(sentence)
+    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
+    loss = model(tensor_input, lm_labels=tensor_input)
+    return math.exp(loss) / len(tokenize_input)     # normalize perplexity by number of tokens
+
+
+def identify_verbs_and_auxiliaries(s):
+    """
+    :param s: input sentence
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:86]: Does this have any normalization applied? I'm thinking lowercasing since the SUBJECT_PRONOUNS are "he" and "she"

File: neutral_generation/smart_convert.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+    :return: dictionary of verbs and list of auxiliaries (if any) with 'she' or 'he' as their subject
+    """
+    SUBJECT_PRONOUNS = ['she', 'he']
+    doc = nlp(s)
+
+    # identify all verbs
+    verbs = set()
+    # this deals with repeating verbs, e.g. "He sings and sings."
+    #   because verb with same text will have different position (makes them unique)
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:95]: nit: extra space

File: neutral_generation/smart_convert.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+        is_conjunct = (
+            possible_conjunct.dep == conj and           # current token is a conjunct
+            # removing the line below, can have a conjunct with an auxiliary verb, e.g. "He was angry and ran away."
+            # possible_conjunct.head.pos == VERB and      # head of current token is a verb
+            possible_conjunct.head in verbs             # the subject of that verb is she / he
+        )
+        if is_conjunct:
+            verbs.add(possible_conjunct)
+
+    # print('verbs: ', verbs)
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:119]: Best to remove debugging lines before committing.

File: neutral_generation/smart_convert.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+
+    auxiliaries = dict()
+    for verb in verbs:
+        auxiliaries[verb] = list()
+    for possible_aux in doc:
+        is_auxiliary = (
+            # current token is an auxiliary verb or negation (not)
+            (possible_aux.dep == aux or possible_aux.dep == neg) and
+            possible_aux.head.pos == VERB and       # head of current token is a verb
+            possible_aux.head in verbs              # the subject of that verb is she / he
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:129]: Does this work in the case of conjuncts?

File: neutral_generation/smart_convert.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+def pluralize_verbs(s, auxiliaries):
+    """
+    :param s: input sentence
+    :param auxiliaries: dictionary of verbs and list of auxiliaries (if any) to be pluralized
+    :return: sentence with plural form of verbs
+    """
+
+    # do_not_change includes 7 tenses in total:
+    #   the 4 future tenses, the 2 past perfect tenses, and the past simple
+    do_not_change = ['past simple', 'past perfect', 'future']
```
Comment (kelliemwebster) [inline neutral_generation/smart_convert.py:149]: Optional.

As discussed offline, this could actually be simplified by only attempting to pluralize the "finite" verb -- the root verb when there are no auxiliaries, else the first auxiliary.

**Review by kelliemwebster — CHANGES_REQUESTED** [review]
Nice code -- neat and well structured.

File: neutral_generation/smart_convert.py (hunk 3 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:20]: Got it. Will do.

File: neutral_generation/constants.py (hunk 2 above)
Comment (tonysun9) [reply neutral_generation/constants.py:60]: Ah, nice catch. I had failed to consider this option.

File: neutral_generation/smart_convert.py (hunk 8 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:159]: Very interesting! Thanks for sharing this tidbit. I'll replace string concatenation with ''.join(...) from now on.

File: neutral_generation/smart_convert.py (hunk 7 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:55]: Good to know. I'll remove this function.

File: neutral_generation/smart_convert.py (hunk 6 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:41]: Got it. Will do.

File: neutral_generation/constants.py (hunk 9 above)
Comment (tonysun9) [reply neutral_generation/constants.py:2]: I see! This is a great observation. I will look into implementing this. 

I think using the dependency structure will also improve runtime significantly since loading and using the LM is what takes the most time in our Python script.

File: neutral_generation/constants.py (hunk 10 above)
Comment (tonysun9) [reply neutral_generation/constants.py:64]: Yes, you're right. SpaCy does tokenize "doesn't" as "does" + "n't". I'll remove the "doesn" --> "don" mapping since the token "doesn" should not appear in our dependency structure.

File: neutral_generation/smart_convert.py (hunk 14 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:119]: Got it. Will do.

File: neutral_generation/smart_convert.py (hunk 13 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:95]: Got it. Will fix.

File: neutral_generation/smart_convert.py (hunk 16 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:149]: Yes, that's a great suggestion. I think that would be a cleaner way of pluralizing verbs.

File: neutral_generation/smart_convert.py (hunk 11 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:48]: Ah, thanks for pointing this out. I'll look into these POS tags.

File: neutral_generation/smart_convert.py (hunk 12 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:86]: I believe I lowercase when checking for potential_subject, but I will do some testing to make sure that this captures our expected cases. Will also make this clearer in the code.

File: neutral_generation/smart_convert.py (hunk 15 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:129]: Thanks for pointing this out. I've added in some checks to make sure that we capture the auxiliaries for conjucts.

File: neutral_generation/smart_convert.py (hunk 5 above)
Comment (tonysun9) [reply neutral_generation/smart_convert.py:32]: Yes, that's a good suggestion. Done.

**Review by melvinjosej — APPROVED** [review]
Thanks Tony. Let's merge these changes.
